bip-hashrate-escrows.mediawiki

```diff
@@ -0,0 +1,469 @@
+
+===== 2. New Messages =====
+
+* M1. "Propose New Escrow"
+* M2. "ACK Escrow Proposal"
+* M3. "Propose Withdrawal"
+* M4. (implied) "ACK Withdrawal"
+* M5. "Execute Deposit"   -- a transfer of BTC from-main-to-side
+* M6. "Execute Withdrawal" -- a transfer of BTC from-side-to-main
+
+
+==== On the Resource Requirements of New Databases ====
+
+The "new" databases are simply reinterpretations of data that are already contained elsewhere in the blockchain. Specifically, M1 M2 and M3 are all located in the block's coinbase txn, and M5 and M6 might be found in any regular txn. M4 is a special case and does not actually need to be included anywhere, so it is not. If you like, you can imagine that the M4s reside in an optional extension block.
+
+In other words, we just rearrange what is already there. Because of this, even though "new databases" are created and stored in memory, the existing bandwidth and storage limits are respected (although, see "M4" below).
+
+
+
+
+=== Adding Sidechains and Tracking Them (D1, M1, M2) ===
+
+==== D1 -- "Escrow_DB" ====
+
+The table below enumerates the new database fields, their size in bytes, and their purpose. In general, an escrow designer (for example, a sidechain-designer), is free to choose any value for these.
+
+Note: Fields 6 through 9 have been intentionally removed. Previously, this section allowed miners to set and commit to voting/waiting periods. However, I have since standardized the periods: withdrawals expire after 6 months (26298 blocks), and they succeed if they ever achieve an ACK score of 13140 or higher. I have removed the waiting period, because anyone who adopts a policy of ignoring all withdrawals with fewer than 400 ACKs will automatically gain all of the benefits of the waiting period. The justification for this change is that it strongly implies that an attack on any one sidechain is an attack on all of them (in a sense, this change makes the "victimhood" of each sidechain "fungible").
+
+{| class="wikitable"
+! Field No.
+! Label
+! Bytes
+! Type
+! Description / Purpose
+|-
+| 1
+| Escrow Number
+| 1
+| uInt
+| A number assigned to the entire escrow. Used to make it easy to refer to each escrow.
+|-
+| 2
+| Active*
+| 2
+| uInt
+| This counter starts at zero when a new entry is added, and increments as a new entry is supported by "M2". The entry either reaches the max value of 4032 (at which point the counter freezes), or else the entry is removed. This enforces the requirement "a soft fork is required to add a new escrow".
+|-
+| 3
+| Escrow Name/Description
+| 120
+| string
+| A human-readable name and description of the sidechain. More than enough space to hold a 32 byte hash. Helps prevent destructive interference among sidechains (see below).
+|-
+| 4
+| Critical Private Key
+| 32
+| hex
+| This is the Double-SHA256 of the binary release of the reference software. It ties the sidechain to a specific software release, and doubles as a way of signing withdrawals (see below).
```

Comment:
What architecture? Compiled with what options? Why is there such a thing as reference software in the first place?

I don't think this field makes sense...

Reply:
It is helpful in cases where there is a disagreement over the definition of a sidechain. It prevents users from being misled over which software to download (if, for example, they want to run "the zcash sidechain" and someone is pretending to offer that software for download).

If a sidechain software has one "reference software", then it inherits the institution of the soft fork. This is great for users who like soft forks. And it is great for developers who want to resist being coerced into hardforking-in bad features into their sidechains.

Reply:
@psztorc  You did not address the "What architecture? Compiled with what options?" part. Is it supposed to be a Linux/Mac/Windows binary...?

Reply:
@pennersr I actually don't see how it matters. The options/configuration instructions could be defined in the source code itself.

More important to me, though, is that the sidechains have a "freedom to fail" design. If so, the sidechain designers are free to make whatever crazy mistakes/choices they think their users want.

But it is important to have a reference implementation, is it not? For exactly the same reasons that Satoshi / Peter Todd are against alternative implementations of Bitcoin -- if they ever disagree over a single byte, there will be consensus failure.

This was just a simple idea on how to try to prevent that.